### ISSUE TITLE
Bring back the .fablemap file to precompile libraries

### ISSUE THREAD

**Issue by alfonsogarciacaro** [fable-compiler/Fable#1257]
See https://github.com/fable-compiler/Fable/blob/v0.7.50/docs/source/blog/Introducing-0-7.md#references

Precompilation of libraries was discarded to allow the final user generate JS code according to their needs (mainly the module system: AMD, commonjs or ES6) and to prevent conflicts among Fable versions. Now that Fable is stabilizing and Webpack supports ES6 is a good moment to bring back the `.fablemap` file back. I don't think libraries should be distributed in precompiled form but there would be two main use cases:

- Fable REPL: it's slower than the CLI and currently doesn't support multiple files so adding libraries is precompiled. However, it should be possible to add a .dll reference and precompiled JS files. @MangelMaxime @ncave 
- Speed up first compilation: Fable could precompile and cache the sources of the libraries. This could include a list with the name of libraries as well the Fable compiler version used so whenever any of this changes, the cache would be invalidated and the libraries precompiled again.

**Comment by MangelMaxime** [comment]
Would be awesome to have it back for the REPL :)

As for now, it can only works with pure bindings which is already opening some interesting usage :)

**Comment by ncave** [comment]
@alfonsogarciacaro @MangelMaxime
IMO it's probably a good idea to first describe what the REPL's multi-file use case would look like (in a bit more detail), so we can see how to make it work, before we start optimizing it.

I'm not saying it's not a good idea, but perhaps there will be multiple ways to achieve the same goal.

**Comment by ncave** [comment]
@alfonsogarciacaro The second reason (use case) above about first compilation sounds like an internal optimization for the Fable compiler, but I'm not sure how it improves over the the timestamp optimization.
I'm probably missing the point somewhere.

**Comment by MangelMaxime** [comment]
I agree with @ncave on both point now 😁.

For discussion about the repl multi file support we already have an opened issue I believe and can speak about it there

**Comment by alfonsogarciacaro** [comment]
@ncave Sorry I wasn't clear in the description. I actually think Fable libraries should **not** include precompiled JS files in the distributed package, this should be done by the final user in the first compilation to speed up the next ones. About other possible optimizations in the past we tried [caching the JS for files that hadn't changed](https://github.com/fable-compiler/Fable/pull/974) but after solving a bottleneck this was not much faster than normal compilation and had other issues so I removed it. Most of the time in the first compilation is [spent on these two lines](https://github.com/fable-compiler/Fable/blob/c02b7177ab7f6c8692c677272fb809d5e761f68d/src/dotnet/dotnet-fable/StateUtil.fs#L102-L103) and I have noticed that FCS parses the project faster with .dll references than with multiple source files.

Anyways, optimization was not the driver for this but adding libraries to the REPL. And the main reason I thought of this is as a possible solution it's because it was already done so it shouldn't be too difficult to bring it back... But I just remember inline expressions are going to be a problem (it wasn't possible to use `inline` in Fable 0.7 libraries and now we use it a lot) so I'm not sure know 😕

**Comment by michaelsg** [comment]
Now that VS 2017 (15.5) is out and can read and manipulate dotnet core-style .fsproj files, the `<Content ...` line to pull in *.fs and *.fsproj interacts with VS in undesirable ways.
1. The content element causes VS to display all files of the project in alphabetical order despite being in the correct order in the `<Compile ...` elements.
2. Manipulating a new-style .fsproj with VS 15.5 removes the `PackagePath="fable"` attribute from the `<Content ...` element.

While you are rethinking library projects:
Is there any other way to get `dotnet pack` to work without messing with VS 15.5?

**Comment by MangelMaxime** [comment]
For me it's a problem of VS 2017 (15.5) it's should not modify the fsproj content.

It should not modify custom lines of the project file. Thus lines are valid, so it should not be changed.

**Comment by alfonsogarciacaro** [comment]
But this only affects libraries, not final user projects? Or are you building a library @michaelsg? If that's the case, maybe using some conditional? (not sure about the syntax in the sample)

```xml
  <ItemGroup Condition="Configuration === 'Packaging'">
    <Content Include="*.fsproj; *.fs" PackagePath="fable\" />
  </ItemGroup>
```

Maybe @enricosada has any suggestion?

**Comment by michaelsg** [comment]
Yes, this only affects library projects. 

Unfortunately, using Packaging configurations doesn't prevent the files from displaying in alphabetical order, even when a non-Packaging configuration is chosen. 

Aren't we abusing the `<Content>` element somewhat by using it to include source and project files? It's been around for a long time and it's meant for including non-compiled files into the project output.

How about something that works everywhere - nuget package creation that does not use `dotnet pack`.
`dotnet fable pack` ? - creates a nuget package with compiled output and Fable project+sources.

**Comment by enricosada** [comment]
> The content element causes VS to display all files of the project in alphabetical order despite being in the correct order in the <Compile ... elements.

that's a know bug of VS afaik

> Unfortunately, using Packaging configurations doesn't prevent the files from displaying in alphabetical order, even when a non-Packaging configuration is chosen.

the issue is you see the `Content` items?

add 

```xml
<Visible>False</Visible>
```

like

```xml
<ItemGroup>
    <Content Include="*.fsproj; *.fs" PackagePath="fable\">
       <Visible>False</Visible>
    </Content>
  </ItemGroup>
```

**Comment by MangelMaxime** [comment]
@enricosada I don't have small repo but you can probably use any of this repo:

- https://github.com/fable-compiler/fable-powerpack/
- https://github.com/MangelMaxime/Fulma/

Fable.Powerpack should be enough IMO.
Fulma is a little more complex because it's use different project with `ProjectReference` that's why I linked it.

**Comment by michaelsg** [comment]
Tried `<Visible>false</Visible>`, but that just hides the fsproj file and still shows sources in alphabetical order. Getting closer :)

**Comment by enricosada** [comment]
Thx i'll try tomorrow

**Comment by michaelsg** [comment]
Just tried moving the content item group before the compile item group. The files appear in compilation order now in the project tree in VS.

```xml
  <ItemGroup>
    <Content Include="*.fs; *.fsproj" PackagePath="fable\">
      <Visible>false</Visible>
    </Content>
  </ItemGroup>
  <ItemGroup>
    <Compile Include="MicroKernel.fs" />
     ...
  </ItemGroup>
```
